File: apis/src/main/java/com/beat/apis/ticket/application/dto/TicketDetail.java
package com.beat.apis.ticket.application.dto;

import java.time.LocalDateTime;
import java.util.Objects;

import com.beat.apis.booking.application.dto.BookingStatusType;

public record TicketDetail(
	Long bookingId,
	String bookerName,
	String bookerPhoneNumber,
	Long scheduleId,
	int purchaseTicketCount,
	LocalDateTime createdAt,
	BookingStatusType bookingStatus,
	String scheduleNumber,
	String bankName,
	String accountNumber,
	String accountHolder
) {
	public static TicketDetail of(
		Long bookingId,
		String bookerName,
		String bookerPhoneNumber,
		Long scheduleId,
		int purchaseTicketCount,
		LocalDateTime createdAt,
		BookingStatusType bookingStatus,
		String scheduleNumber,
		String bankName,
		String accountNumber,
		String accountHolder) {
		return new TicketDetail(
			bookingId,
			bookerName,
			bookerPhoneNumber,
			scheduleId,
			purchaseTicketCount,
			createdAt,
			bookingStatus,
			scheduleNumber,
			Objects.requireNonNullElse(bankName, ""),
			Objects.requireNonNullElse(accountNumber, ""),
			Objects.requireNonNullElse(accountHolder, "")
		);
	}
}

File: apis/src/main/kotlin/com/beat/apis/ticket/api/response/TicketDetail.kt
package com.beat.apis.ticket.api.response

import com.beat.apis.booking.api.type.BookingStatusType
import com.beat.apis.ticket.application.result.TicketDetailResult
import java.time.LocalDateTime

@ConsistentCopyVisibility
data class TicketDetail private constructor(
    val bookingId: Long?,
    val bookerName: String?,
    val bookerPhoneNumber: String?,
    val scheduleId: Long?,
    val purchaseTicketCount: Int,
    val createdAt: LocalDateTime?,
    val bookingStatus: BookingStatusType?,
    val scheduleNumber: String?,
    val bankName: String,
    val accountNumber: String,
    val accountHolder: String,
    val deletable: Boolean,
) {
    companion object {
        fun from(result: TicketDetailResult): TicketDetail = TicketDetail(
            bookingId = result.bookingId,
            bookerName = result.bookerName,
            bookerPhoneNumber = result.bookerPhoneNumber,
            scheduleId = result.scheduleId,
            purchaseTicketCount = result.purchaseTicketCount,
            createdAt = result.createdAt,
            bookingStatus = result.bookingStatus?.let(BookingStatusType::valueOf),
            scheduleNumber = result.scheduleNumber,
            bankName = result.bankName,
            accountNumber = result.accountNumber,
            accountHolder = result.accountHolder,
            deletable = result.deletable,
        )
    }
}
